### ISSUE TITLE
AttributeError: 'datetime.datetime' object has no attribute 'times'

### ISSUE THREAD

**Issue by lucbouge** [python/cpython#120788]
# Bug report

### Bug description:

I want to use Python 3.12.4 on MacOS Intel 14.5 (23F79).
I launched the installation with Homebrew: `brew install python@3.12`
The post-installation phase failed in the `/usr/local/Cellar/python@3.12/3.12.4/bin/python3.12 -Im ensurepip` phase
After some digging, I found that it was caused by a failure in accessing the `strftime` method of a `datetime.datetime` object in the `/usr/local/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/xmlrpc/client.py` library, line 269.
```
    try:
        return _day0.strftime(fmt) == '0001'
    except ValueError:
        return False
```
Object `_day0` has type `datetime.datetime` and calling method `strftime` fails with message `AttributeError: 'datetime.datetime' object has no attribute 'times'`

Hopefully, I could comment out this line, so that I could finally complete the post-installation phase.

However, once everything was fine, I tested the following code, and the same message showed up again.
```
from datetime import datetime
dt = datetime.now()
s = dt.strftime("%Y")
print(s)
```

```
% python test.py 
Traceback (most recent call last):
  File "/Users/bouge/Repositories/no_save/tmp/test.py", line 4, in <module>
    s = dt.strftime("%Y")
        ^^^^^^^^^^^^^^^^^
AttributeError: 'datetime.datetime' object has no attribute 'times'. Did you mean: 'time'?
```

Therefore, I suspect there is a bug in the datetime.datetime library, probably a typo?

### CPython versions tested on:

3.12

### Operating systems tested on:

macOS

**Comment by lucbouge** [comment]
NB. The error also prevents `pdm install` from completing correctly.

```
% pdm install
Synchronizing working set with resolved packages: 15 to add, 0 to update, 0 to remove

  ✖ Install attrs 23.2.0 failed
  ✖ Install typing-extensions 4.10.0 failed
  ✖ Install tzdata 2024.1 failed
  ✖ Install types-pytz 2024.1.0.20240203 failed
  ✖ Install six 1.16.0 failed
  ✖ Install typeguard 4.3.0 failed
  ✖ Install imap-tools 1.6.0 failed
  ✖ Install pytz 2024.1 failed
  ✖ Install unidecode 1.3.8 failed
  ✖ Install python-dateutil 2.9.0.post0 failed
  ✖ Install pandas-stubs 2.2.0.240218 failed
  ✖ Install more-itertools 10.3.0 failed
  ✖ Install python-libraries 0.1.0 failed
  ✖ Install pandas 2.2.2 failed
  ✖ Install numpy 1.26.4 failed
Retry failed jobs
  ✖ Install attrs 23.2.0 failed
  ✖ Install typing-extensions 4.10.0 failed
  ✖ Install six 1.16.0 failed
  ✖ Install python-dateutil 2.9.0.post0 failed
  ✖ Install tzdata 2024.1 failed
  ✖ Install typeguard 4.3.0 failed
  ✖ Install types-pytz 2024.1.0.20240203 failed
  ✖ Install imap-tools 1.6.0 failed
  ✖ Install more-itertools 10.3.0 failed
  ✖ Install unidecode 1.3.8 failed
  ✖ Install pytz 2024.1 failed
  ✖ Install pandas-stubs 2.2.0.240218 failed
  ✖ Install pandas 2.2.2 failed
  ✖ Install python-libraries 0.1.0 failed
  ✖ Install numpy 1.26.4 failed

ERRORS:
add attrs failed:
Traceback (most recent call last):
  File 
"/usr/local/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/pdm/installers/synchronizers.py", line 282, 
in install_candidate
    self.manager.install(can)
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/pdm/installers/manager.py", line 33, in 
install
    prepared.build(),
    ^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/pdm/models/candidates.py", line 404, in build
    self._obtain(allow_all=False)
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/pdm/models/candidates.py", line 440, in 
_obtain
    self.link = _find_best_match_link(
                ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/pdm/models/candidates.py", line 107, in 
_find_best_match_link
    found = attempt_to_find()
            ^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/pdm/models/candidates.py", line 98, in 
attempt_to_find
    best = finder.find_best_match(req.as_line(), hashes=hashes).best
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/unearth/finder.py", line 397, in 
find_best_match
    best_match = next(iter(applicable_candidates), None)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/unearth/finder.py", line 330, in 
_find_packages_from_requirement
    yield from self._find_packages(requirement.name, allow_yanked)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/unearth/finder.py", line 296, in 
_find_packages
    return sorted(all_packages, key=self._sort_key, reverse=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/unearth/collector.py", line 182, in 
collect_links_from_location
    yield from _collect_links_from_index(session, location)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/unearth/collector.py", line 205, in 
_collect_links_from_index
    page = fetch_page(session, location, headers)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/unearth/collector.py", line 190, in 
fetch_page
    resp = _get_html_response(session, location, headers)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/unearth/collector.py", line 229, in 
_get_html_response
    resp = session.get(
           ^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/httpx/_client.py", line 1054, in get
    return self.request(
           ^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/httpx/_client.py", line 827, in request
    return self.send(request, auth=auth, follow_redirects=follow_redirects)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/httpx/_client.py", line 914, in send
    response = self._send_handling_auth(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/httpx/_client.py", line 942, in 
_send_handling_auth
    response = self._send_handling_redirects(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/httpx/_client.py", line 979, in 
_send_handling_redirects
    response = self._send_single_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/httpx/_client.py", line 1015, in 
_send_single_request
    response = transport.handle_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/hishel/_sync/_transports.py", line 209, in 
handle_request
    self._storage.store(
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/hishel/_sync/_storages.py", line 136, in 
store
    self._serializer.dumps(response=response, request=request, metadata=metadata),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/pdm/2.15.4/libexec/lib/python3.12/site-packages/pdm/models/session.py", line 83, in dumps
    "created_at": metadata["created_at"].strftime(self.DATETIME_FORMAT),
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'datetime.datetime' object has no attribute 'times'. Did you mean: 'time'?

[...]

See /Users/bouge/Library/Logs/pdm/pdm-install-hwlxbc1v.log for detailed debug log.
[InstallationError]: Some package operations are not complete yet
WARNING: Add '-v' to see the detailed traceback

```

**Comment by lucbouge** [comment]
Looking precisely at the execution trace (`python -vvv`), I realized that `datetime` was loaded from a dynamic library called `_datetime.cpython-312-darwin.so`.

Unfortunately, such a library existed in `/usr/local/lib/python3.12/lib-dynload`, which hid the one in `/usr/local/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/lib-dynload`.

This spurious library was probably generated by a previous attempt to directly compile the python3.12 source tarball.

Removing the entire `/usr/local/lib/python3.12/lib-dynload` directory solved my problem.

**Comment by RichardFevrier** [comment]
Annoying when some people like me are using an immutable distro...

**Comment by philippfreyer** [comment]
Fix did not work for me...

**Comment by christophlsa** [comment]
fyi: I had a similar error (different attribute and different module but it was also trying `%Y`):

```
Traceback (most recent call last):
  File "/usr/bin/supervisord", line 5, in <module>
    from supervisor.supervisord import main
  File "/usr/lib/python3.12/site-packages/supervisor/supervisord.py", line 38, in <module>
    from supervisor.medusa import asyncore_25 as asyncore
  File "/usr/lib/python3.12/site-packages/supervisor/medusa/asyncore_25.py", line 58, in <module>
    from supervisor.compat import as_string, as_bytes
  File "/usr/lib/python3.12/site-packages/supervisor/compat.py", line 67, in <module>
    import xmlrpc.client as xmlrpclib
  File "/usr/lib/python3.12/xmlrpc/client.py", line 272, in <module>
    if _try('%Y'):      # Mac OS X
       ^^^^^^^^^^
  File "/usr/lib/python3.12/xmlrpc/client.py", line 269, in _try
    return _day0.strftime(fmt) == '0001'
           ^^^^^^^^^^^^^^^^^^^
AttributeError: 'datetime.datetime' object has no attribute 'task'
```
This error was introduced with 3.12.6 (3.12.5 was fine) and was fixed for me in 3.12.7.
